cmd/snap-update-ns/utils.go

```diff
@@ -336,7 +336,7 @@ func planWritableMimic(dir string) ([]*Change, error) {
 			// parsing the mount table and exploring what is really
 			// there and this is not how the undo logic is
 			// designed.
-			Name: dir, Dir: safeKeepingDir, Options: []string{"bind"}},
+			Name: dir, Dir: safeKeepingDir, Options: []string{"rbind"}},
```

Comment:
This change contradicts the preceding comment.

I guess the idea is that if you lazy unmount then you don't need to 'parse the mount table and explore what is really there'. I do wonder though, if you lazy unmount everything, what guarantee is there that something is missed and left hanging around in the mount table? Do the tests adequately cover this case?

Also, What triggered this? You mentioned that PR #4644 now passes with this change, but looking at that, I don't see where we have things mounted under bind mounts (indeed, a requirement for layouts is precisely that we do *not* allow specifying a layout that is under another. Does this have to do with writable-paths and /etc? If so, in the above comment change, it might be worth mentioning this use case.

Reply:
I will respond in full later but this was triggered by using layouts on core where writable paths system puts various bind mounts all over the filesystem. In particular this fixes the spread layout test in another PR of mine. One example was /var/lib/ which is a zoo of things that only worked because writable paths added a lot of things to make that read-only place writable.

If you unbind a tree of mount points the kernel handles unmounting those in the right order. We just take advantage of that.